{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Python-dotenv could not parse statement starting at line 1\n",
      "Python-dotenv could not parse statement starting at line 7\n"
     ]
    }
   ],
   "source": [
    "# Import necessary libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from datasets import Dataset\n",
    "from ragas.metrics import AnswerSimilarity, AnswerCorrectness\n",
    "from ragas.evaluation import evaluate\n",
    "import os\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "os.environ['OPENAI_API_KEY'] = os.getenv(\"OPENAI_API_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the CSV file\n",
    "csv_file_path = 'C:/Users/York Yong/OneDrive - Singapore Management University/Desktop/Pokemon Game RAG Chatbot/Basline LLM evaluation/Eval_ChatGPT4o.csv'\n",
    "\n",
    "try:\n",
    "    data = pd.read_csv(csv_file_path, encoding='utf-8')\n",
    "except UnicodeDecodeError:\n",
    "    data = pd.read_csv(csv_file_path, encoding='latin1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "561b898895ef4269b07f09c555ee6b7a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8016dc0b214d4e49bb55504dce6267a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4f45e5e152f74b1395b90f299135a5ed",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "455ce54c33ce4b76a2a51088a8e2a2d6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "381f2a6a56c14ea9b8937ccf702fd50d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c93e261bbda84bf69fc4f42dc64f0b1c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4e184c30e30e4196955578bc25a65d91",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e63dc7a944b1471989451c77e435ed12",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "820414d58ed54356b638af848370c66a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7bb1c1aae26948b1978c27d00c2d765d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "73089ee791d14a87bbb31225f5699e87",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "aa210fd2c4af4a0091669f022adffc7b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "106da0707d2c47d2981f73b911b5b182",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "933f87ff483c4ed7892c37a7ffb05d0c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Failed to parse output. Returning None.\n",
      "C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\ragas\\evaluation.py:299: RuntimeWarning: Mean of empty slice\n",
      "  value = np.nanmean(self.scores[cn])\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dbd9394ffd5b41fba34191d900ad5f00",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "04448b9fed2d4b2c86619323e8d37d96",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "567dc06d348f43ce8c74d10ac418b6ad",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ac1b38edb3dc4d46abb76e999eada9f3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "63907c03e80b417793de8d5b73ccb230",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e85cacb9747e4f82ab08c454182504eb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a5d9b7b5d7cb40e4b45ef6cf2001752a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "720522b14a30438a954d479ab4c97c2b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1afc1e16bb2949b8afbe254e27b93b03",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "22fa1891c38d44abafa088cb6442429f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d15e3f0565ca4b14843e94d338b0e9bf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "95170d5696b941d5805036da62034103",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d7da2d60e7dc428ea3665dfe79b94035",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8e5a0c00b67745548788bbb74c06f155",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a0547abb3df345fa987dbab84a305702",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bc73705740e749909cd64dc238edd93d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "615f10e234df40e38eecd0588fea418a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f52ce553a1f14186b9170090079f91ca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f647c89bfa134c95aad32aebea9ce52f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "35562a7664ab4c34aaddafda163b65db",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "494c8961b4354c54b79d89abc919d932",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "54b635d560404c02b661598c38300082",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "041d448b5fee467d923386fd1c33188a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f08fb42f44784dc6ae75a50283257a82",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ef4561490f064f0896efdad27167f2bd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "162bb50383b24b379a9fc64d3fb45152",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c5c951ebfae34c0c8e1ce2f5a9d831dd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6c450c268d6748bfb55d647d05f917ea",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "030e6678c9e84036b5583fb7431dff52",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e364dedd4c764822a1360ff81e5fa639",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c335c3f3c8c2402391128d57c43e6fca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "67e9deb018fb45efb0267a0671db2ca4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "33f54f8b082947ee83f2df7f962889bd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "16d768b0ce92412b99724a12e6869fc0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e4dce71102b34283acfd9f38586438f2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Failed to parse output. Returning None.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9b9ea402e5c64f069032721a06f1ad9d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-101' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-102' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-160' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-161' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-219' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-220' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-278' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-279' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-340' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-341' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-399' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-400' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-458' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-459' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-517' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-518' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-576' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-577' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-639' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-640' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-698' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-699' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-757' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-758' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-816' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-817' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-875' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-876' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-934' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-935' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-997' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-998' coro=<AsyncClient.aclose() done, defined at C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\tasks.py\", line 232, in __step\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"C:\\Users\\York Yong\\AppData\\Roaming\\Python\\Python310\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1181, in aclose\n",
      "    self._transport.close()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\proactor_events.py\", line 109, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 753, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\asyncio\\base_events.py\", line 515, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n"
     ]
    }
   ],
   "source": [
    "# Define the metrics to be evaluated\n",
    "metrics = [\n",
    "    AnswerSimilarity(), \n",
    "    AnswerCorrectness()\n",
    "]\n",
    "\n",
    "# Initialize columns for results\n",
    "data['answer_similarity'] = None\n",
    "data['answer_correctness'] = None\n",
    "\n",
    "# Iterate over each row to evaluate the metrics individually\n",
    "for idx, row in data.iterrows():\n",
    "    evaluation_data = Dataset.from_dict({\n",
    "        'question': [row['question']],\n",
    "        'ground_truth': [row['ground_truth']],\n",
    "        'answer': [row['answer']]\n",
    "    })\n",
    "    \n",
    "    evaluation_result = evaluate(evaluation_data, metrics=metrics)\n",
    "    \n",
    "    # Store results back in the DataFrame\n",
    "    data.at[idx, 'answer_similarity'] = evaluation_result['answer_similarity']\n",
    "    data.at[idx, 'answer_correctness'] = evaluation_result['answer_correctness']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>question</th>\n",
       "      <th>ground_truth</th>\n",
       "      <th>answer</th>\n",
       "      <th>answer_similarity</th>\n",
       "      <th>answer_correctness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>What are the starting choices for Pokémon in P...</td>\n",
       "      <td>The starter Pokémon options in Paldea are Spri...</td>\n",
       "      <td>The starting choices for Pokémon in Paldea are...</td>\n",
       "      <td>0.957912</td>\n",
       "      <td>0.689478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>What special mounts are available in Pokémon S...</td>\n",
       "      <td>Scarlet players can catch and ride Koraidon, w...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the special mou...</td>\n",
       "      <td>0.908454</td>\n",
       "      <td>0.227113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>How does fast travel work in the game?</td>\n",
       "      <td>Fast travel is unlocked immediately and can be...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, fast travel all...</td>\n",
       "      <td>0.865435</td>\n",
       "      <td>0.489086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>How do you initiate battles with other trainers?</td>\n",
       "      <td>Unlike past games, you must talk to trainers t...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, initiating batt...</td>\n",
       "      <td>0.914648</td>\n",
       "      <td>0.365026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>What can you collect from gimmighoul chests?</td>\n",
       "      <td>You can collect coins from gimmighoul chests.</td>\n",
       "      <td>In Pokémon Scarlet and Violet, you can collect...</td>\n",
       "      <td>0.920761</td>\n",
       "      <td>0.922498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>How are items indicated in the game world?</td>\n",
       "      <td>Items are indicated by pokeballs or shiny spot...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, items in the ga...</td>\n",
       "      <td>0.885433</td>\n",
       "      <td>0.682897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>What benefits do sandwiches offer in the game?</td>\n",
       "      <td>Sandwiches provide buffs for 30 minutes durati...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, sandwiches prov...</td>\n",
       "      <td>0.882245</td>\n",
       "      <td>0.220561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>How can you prevent getting lost using the map?</td>\n",
       "      <td>Press the right analog stick to fix the map or...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, preventing gett...</td>\n",
       "      <td>0.768602</td>\n",
       "      <td>0.350045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>What is a recommended tactic for leveling up q...</td>\n",
       "      <td>Eating a Ham Sandwich (#80) to increase normal...</td>\n",
       "      <td>A recommended tactic for leveling up quickly i...</td>\n",
       "      <td>0.782101</td>\n",
       "      <td>0.195525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>What are the three main storylines in Pokémon ...</td>\n",
       "      <td>The Gym Challenge, Starfall Street, and the Pa...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the three main ...</td>\n",
       "      <td>0.862764</td>\n",
       "      <td>0.429977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>What abilities do defeating Titan Pokémon unlock?</td>\n",
       "      <td>Defeating Titan Pokémon unlocks new abilities ...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, defeating Titan...</td>\n",
       "      <td>0.905518</td>\n",
       "      <td>0.687918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>What are the recommended levels for fighting t...</td>\n",
       "      <td>The recommended levels for fighting Klawf are ...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the recommended...</td>\n",
       "      <td>0.865989</td>\n",
       "      <td>0.966497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>How does the battle mechanic change with gym l...</td>\n",
       "      <td>The final Pokémon used by gym leaders will Ter...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the battle mech...</td>\n",
       "      <td>0.870893</td>\n",
       "      <td>0.448492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>What is the challenge format for Starfall Stre...</td>\n",
       "      <td>You must defeat 30 Pokémon in autobattles befo...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the Starfall St...</td>\n",
       "      <td>0.810889</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>How do you capture a second Koraidon or Miraid...</td>\n",
       "      <td>After completing the main story, a second Kora...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, capturing a sec...</td>\n",
       "      <td>0.891147</td>\n",
       "      <td>0.904605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>How do you breed Pokémon in Scarlet and Violet?</td>\n",
       "      <td>Breeding involves having a Ditto and another P...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, breeding Pokémo...</td>\n",
       "      <td>0.908255</td>\n",
       "      <td>0.709207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>What is the significance of the \"Area Zero\" mi...</td>\n",
       "      <td>It's a deep exploration mission involving the ...</td>\n",
       "      <td>The \"Area Zero\" mission in Pokémon Scarlet and...</td>\n",
       "      <td>0.809698</td>\n",
       "      <td>0.923578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>What is required to evolve Pokémon like Eevee ...</td>\n",
       "      <td>Eevee requires specific stones or conditions l...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, evolving Eevee ...</td>\n",
       "      <td>0.878603</td>\n",
       "      <td>0.719651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>How can you earn more LP in the game?</td>\n",
       "      <td>LP can be earned through Tera Raid Battles and...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, you can earn mo...</td>\n",
       "      <td>0.883585</td>\n",
       "      <td>0.336257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>What type of team is recommended for the Elite...</td>\n",
       "      <td>A diverse team covering multiple types such as...</td>\n",
       "      <td>To effectively tackle the Elite Four in Pokémo...</td>\n",
       "      <td>0.865171</td>\n",
       "      <td>0.216293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>How do Tera Raid Battles differ from regular b...</td>\n",
       "      <td>They allow multiple players to fight a single,...</td>\n",
       "      <td>Tera Raid Battles in Pokémon Scarlet and Viole...</td>\n",
       "      <td>0.825973</td>\n",
       "      <td>0.902922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>What is the role of gimmighoul coins?</td>\n",
       "      <td>They are used to evolve Gimmighoul into Gholde...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, Gimmighoul Coin...</td>\n",
       "      <td>0.913784</td>\n",
       "      <td>0.603446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>How do you obtain special items like the Chipp...</td>\n",
       "      <td>These can be found on the ground, bought at au...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, obtaining speci...</td>\n",
       "      <td>0.786776</td>\n",
       "      <td>0.339551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>What is the method to ensure a successful catc...</td>\n",
       "      <td>Turn off auto-save, use a Master Ball or set u...</td>\n",
       "      <td>Catching legendary Pokémon in Pokémon Scarlet ...</td>\n",
       "      <td>0.809653</td>\n",
       "      <td>0.338777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>What pokemon does Montenevera Chost Gym leader...</td>\n",
       "      <td>Rymn, the gym leader for Montenevera Ghost Gym...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the Montenevera...</td>\n",
       "      <td>0.931193</td>\n",
       "      <td>0.76221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>What is the recipe for Electric Shiny Sandwich?</td>\n",
       "      <td>You will need the following:\\nYellow Bell Pepp...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, making a sandwi...</td>\n",
       "      <td>0.821579</td>\n",
       "      <td>0.955395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Who is the Real Cassiopeia?</td>\n",
       "      <td>The Real Cassiopeia is Penny</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the real identi...</td>\n",
       "      <td>0.881536</td>\n",
       "      <td>0.220388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>What are the weakness of the Titan Bombirdier?</td>\n",
       "      <td>Bombirdier is weak to Fairy, Rock and Electric.</td>\n",
       "      <td>The Titan Bombirdier in Pokémon Scarlet and Vi...</td>\n",
       "      <td>0.889183</td>\n",
       "      <td>0.487001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Where can I heal my pokemon in Area Zero?</td>\n",
       "      <td>You can use the bed in the research stations t...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, Area Zero is a ...</td>\n",
       "      <td>0.844975</td>\n",
       "      <td>0.863418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>What moves do I need to battle against the Pro...</td>\n",
       "      <td>You will need good Fairy, Psychic, Ground and ...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, battling agains...</td>\n",
       "      <td>0.828336</td>\n",
       "      <td>0.282084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>What is required to evolve Charcadet into Arma...</td>\n",
       "      <td>Auspicious Armor</td>\n",
       "      <td>In Pokémon Scarlet, evolving Charcadet into Ar...</td>\n",
       "      <td>0.827349</td>\n",
       "      <td>0.206837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Which pokemon requires walking with Let's Go t...</td>\n",
       "      <td>Pawno with walking 1000 steps to Pawmot. Bramb...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, there are a few...</td>\n",
       "      <td>0.851082</td>\n",
       "      <td>0.77527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>In the Champion Assessment, what is the name o...</td>\n",
       "      <td>Tulip - Psychic Type Gym Leader located in Alf...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, during the Cham...</td>\n",
       "      <td>0.865514</td>\n",
       "      <td>0.403871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>How to defeat Elite Four Member Hassel?</td>\n",
       "      <td>Hassel is a Dragon specialist who is weak to I...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, Hassel is one o...</td>\n",
       "      <td>0.901105</td>\n",
       "      <td>0.86164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Where is Pokemon League?</td>\n",
       "      <td>It is located in the back of Mesagoza, in the ...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the Pokémon Lea...</td>\n",
       "      <td>0.855798</td>\n",
       "      <td>0.642521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>How do you turn off the time machine in Area Z...</td>\n",
       "      <td>You have to place the book Arven gave you earl...</td>\n",
       "      <td>To turn off the time machine in Area Zero in P...</td>\n",
       "      <td>0.803323</td>\n",
       "      <td>0.200831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Where can you find gimmighoul's chest?</td>\n",
       "      <td>You can go to the top of the Watchtowers.</td>\n",
       "      <td>In Pokémon Scarlet and Violet, Gimmighoul's Ch...</td>\n",
       "      <td>0.739195</td>\n",
       "      <td>0.184799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Where can I find soothe bells?</td>\n",
       "      <td>It is a hold item purchased at the Delibird st...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the Soothe Bell...</td>\n",
       "      <td>0.813806</td>\n",
       "      <td>0.203451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>I keep going in the wrong direction, how do I ...</td>\n",
       "      <td>When you open your map, you might notice it ca...</td>\n",
       "      <td>If you find yourself going in the wrong direct...</td>\n",
       "      <td>0.839107</td>\n",
       "      <td>0.209777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>I've heard I can climb mountains, how do I do ...</td>\n",
       "      <td>Upon defeating the LAST titan you will be able...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, you can climb m...</td>\n",
       "      <td>0.834199</td>\n",
       "      <td>0.33355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>I can't get past gym 6, help!</td>\n",
       "      <td>Gym #6 - Montenevera Ghost Gym - Rymn. Weaknes...</td>\n",
       "      <td>If you're having trouble getting past Gym 6 in...</td>\n",
       "      <td>0.888818</td>\n",
       "      <td>0.395281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>How do I fight Dragon pokemon?</td>\n",
       "      <td>Dragon is weak to: Ice, Fairy and Dragon Types</td>\n",
       "      <td>Fighting Dragon-type Pokémon can be challengin...</td>\n",
       "      <td>0.865473</td>\n",
       "      <td>0.45321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>What is the final battle mission of the game?</td>\n",
       "      <td>After defeating the professor, thats not all!...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, the final battl...</td>\n",
       "      <td>0.823552</td>\n",
       "      <td>0.205888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>How do I get a Polteageist?</td>\n",
       "      <td>Pokemon who need other items to evolve: Sinist...</td>\n",
       "      <td>To get a Polteageist in Pokémon Scarlet and Vi...</td>\n",
       "      <td>0.923085</td>\n",
       "      <td>0.605771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>I'm trying to get a Marill, how do I do that?</td>\n",
       "      <td>Pokemon that require a high level of friendshi...</td>\n",
       "      <td>To get a Marill in Pokémon Scarlet and Violet,...</td>\n",
       "      <td>0.86395</td>\n",
       "      <td>0.290987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>List down all the gyms.</td>\n",
       "      <td>There are 8 gyms - Cortondo, Artazon, Levinici...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, there are eight...</td>\n",
       "      <td>0.838411</td>\n",
       "      <td>0.209603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>How does the \"Let's Go\" feature affect Pokémon...</td>\n",
       "      <td>Certain Pokémon evolve after walking a require...</td>\n",
       "      <td>The \"Let's Go\" feature in Pokémon Scarlet and ...</td>\n",
       "      <td>0.896058</td>\n",
       "      <td>0.599014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>What happens after I defeat the Elite Four?</td>\n",
       "      <td>After youve defeated Geeta, youre the newest...</td>\n",
       "      <td>After you defeat the Elite Four in Pokémon Sca...</td>\n",
       "      <td>0.832983</td>\n",
       "      <td>0.649428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>What level are the Professor's Pokemon in Area...</td>\n",
       "      <td>Level 66 and Level 67</td>\n",
       "      <td>In Pokémon Scarlet and Violet, during the fina...</td>\n",
       "      <td>0.791244</td>\n",
       "      <td>0.197811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>How do I evolve a Sunflora?</td>\n",
       "      <td>Pokemon who require stones to evolve: Sunkern ...</td>\n",
       "      <td>In Pokémon Scarlet and Violet, Sunflora is the...</td>\n",
       "      <td>0.922454</td>\n",
       "      <td>0.313947</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             question  \\\n",
       "0   What are the starting choices for Pokémon in P...   \n",
       "1   What special mounts are available in Pokémon S...   \n",
       "2              How does fast travel work in the game?   \n",
       "3    How do you initiate battles with other trainers?   \n",
       "4        What can you collect from gimmighoul chests?   \n",
       "5          How are items indicated in the game world?   \n",
       "6      What benefits do sandwiches offer in the game?   \n",
       "7     How can you prevent getting lost using the map?   \n",
       "8   What is a recommended tactic for leveling up q...   \n",
       "9   What are the three main storylines in Pokémon ...   \n",
       "10  What abilities do defeating Titan Pokémon unlock?   \n",
       "11  What are the recommended levels for fighting t...   \n",
       "12  How does the battle mechanic change with gym l...   \n",
       "13  What is the challenge format for Starfall Stre...   \n",
       "14  How do you capture a second Koraidon or Miraid...   \n",
       "15    How do you breed Pokémon in Scarlet and Violet?   \n",
       "16  What is the significance of the \"Area Zero\" mi...   \n",
       "17  What is required to evolve Pokémon like Eevee ...   \n",
       "18              How can you earn more LP in the game?   \n",
       "19  What type of team is recommended for the Elite...   \n",
       "20  How do Tera Raid Battles differ from regular b...   \n",
       "21              What is the role of gimmighoul coins?   \n",
       "22  How do you obtain special items like the Chipp...   \n",
       "23  What is the method to ensure a successful catc...   \n",
       "24  What pokemon does Montenevera Chost Gym leader...   \n",
       "25    What is the recipe for Electric Shiny Sandwich?   \n",
       "26                        Who is the Real Cassiopeia?   \n",
       "27    What are the weakness of the Titan Bombirdier?    \n",
       "28          Where can I heal my pokemon in Area Zero?   \n",
       "29  What moves do I need to battle against the Pro...   \n",
       "30  What is required to evolve Charcadet into Arma...   \n",
       "31  Which pokemon requires walking with Let's Go t...   \n",
       "32  In the Champion Assessment, what is the name o...   \n",
       "33            How to defeat Elite Four Member Hassel?   \n",
       "34                           Where is Pokemon League?   \n",
       "35  How do you turn off the time machine in Area Z...   \n",
       "36             Where can you find gimmighoul's chest?   \n",
       "37                     Where can I find soothe bells?   \n",
       "38  I keep going in the wrong direction, how do I ...   \n",
       "39  I've heard I can climb mountains, how do I do ...   \n",
       "40                      I can't get past gym 6, help!   \n",
       "41                     How do I fight Dragon pokemon?   \n",
       "42      What is the final battle mission of the game?   \n",
       "43                        How do I get a Polteageist?   \n",
       "44      I'm trying to get a Marill, how do I do that?   \n",
       "45                            List down all the gyms.   \n",
       "46  How does the \"Let's Go\" feature affect Pokémon...   \n",
       "47        What happens after I defeat the Elite Four?   \n",
       "48  What level are the Professor's Pokemon in Area...   \n",
       "49                        How do I evolve a Sunflora?   \n",
       "\n",
       "                                         ground_truth  \\\n",
       "0   The starter Pokémon options in Paldea are Spri...   \n",
       "1   Scarlet players can catch and ride Koraidon, w...   \n",
       "2   Fast travel is unlocked immediately and can be...   \n",
       "3   Unlike past games, you must talk to trainers t...   \n",
       "4       You can collect coins from gimmighoul chests.   \n",
       "5   Items are indicated by pokeballs or shiny spot...   \n",
       "6   Sandwiches provide buffs for 30 minutes durati...   \n",
       "7   Press the right analog stick to fix the map or...   \n",
       "8   Eating a Ham Sandwich (#80) to increase normal...   \n",
       "9   The Gym Challenge, Starfall Street, and the Pa...   \n",
       "10  Defeating Titan Pokémon unlocks new abilities ...   \n",
       "11  The recommended levels for fighting Klawf are ...   \n",
       "12  The final Pokémon used by gym leaders will Ter...   \n",
       "13  You must defeat 30 Pokémon in autobattles befo...   \n",
       "14  After completing the main story, a second Kora...   \n",
       "15  Breeding involves having a Ditto and another P...   \n",
       "16  It's a deep exploration mission involving the ...   \n",
       "17  Eevee requires specific stones or conditions l...   \n",
       "18  LP can be earned through Tera Raid Battles and...   \n",
       "19  A diverse team covering multiple types such as...   \n",
       "20  They allow multiple players to fight a single,...   \n",
       "21  They are used to evolve Gimmighoul into Gholde...   \n",
       "22  These can be found on the ground, bought at au...   \n",
       "23  Turn off auto-save, use a Master Ball or set u...   \n",
       "24  Rymn, the gym leader for Montenevera Ghost Gym...   \n",
       "25  You will need the following:\\nYellow Bell Pepp...   \n",
       "26                       The Real Cassiopeia is Penny   \n",
       "27    Bombirdier is weak to Fairy, Rock and Electric.   \n",
       "28  You can use the bed in the research stations t...   \n",
       "29  You will need good Fairy, Psychic, Ground and ...   \n",
       "30                                   Auspicious Armor   \n",
       "31  Pawno with walking 1000 steps to Pawmot. Bramb...   \n",
       "32  Tulip - Psychic Type Gym Leader located in Alf...   \n",
       "33  Hassel is a Dragon specialist who is weak to I...   \n",
       "34  It is located in the back of Mesagoza, in the ...   \n",
       "35  You have to place the book Arven gave you earl...   \n",
       "36          You can go to the top of the Watchtowers.   \n",
       "37  It is a hold item purchased at the Delibird st...   \n",
       "38  When you open your map, you might notice it ca...   \n",
       "39  Upon defeating the LAST titan you will be able...   \n",
       "40  Gym #6 - Montenevera Ghost Gym - Rymn. Weaknes...   \n",
       "41     Dragon is weak to: Ice, Fairy and Dragon Types   \n",
       "42  After defeating the professor, thats not all!...   \n",
       "43  Pokemon who need other items to evolve: Sinist...   \n",
       "44  Pokemon that require a high level of friendshi...   \n",
       "45  There are 8 gyms - Cortondo, Artazon, Levinici...   \n",
       "46  Certain Pokémon evolve after walking a require...   \n",
       "47  After youve defeated Geeta, youre the newest...   \n",
       "48                              Level 66 and Level 67   \n",
       "49  Pokemon who require stones to evolve: Sunkern ...   \n",
       "\n",
       "                                               answer answer_similarity  \\\n",
       "0   The starting choices for Pokémon in Paldea are...          0.957912   \n",
       "1   In Pokémon Scarlet and Violet, the special mou...          0.908454   \n",
       "2   In Pokémon Scarlet and Violet, fast travel all...          0.865435   \n",
       "3   In Pokémon Scarlet and Violet, initiating batt...          0.914648   \n",
       "4   In Pokémon Scarlet and Violet, you can collect...          0.920761   \n",
       "5   In Pokémon Scarlet and Violet, items in the ga...          0.885433   \n",
       "6   In Pokémon Scarlet and Violet, sandwiches prov...          0.882245   \n",
       "7   In Pokémon Scarlet and Violet, preventing gett...          0.768602   \n",
       "8   A recommended tactic for leveling up quickly i...          0.782101   \n",
       "9   In Pokémon Scarlet and Violet, the three main ...          0.862764   \n",
       "10  In Pokémon Scarlet and Violet, defeating Titan...          0.905518   \n",
       "11  In Pokémon Scarlet and Violet, the recommended...          0.865989   \n",
       "12  In Pokémon Scarlet and Violet, the battle mech...          0.870893   \n",
       "13  In Pokémon Scarlet and Violet, the Starfall St...          0.810889   \n",
       "14  In Pokémon Scarlet and Violet, capturing a sec...          0.891147   \n",
       "15  In Pokémon Scarlet and Violet, breeding Pokémo...          0.908255   \n",
       "16  The \"Area Zero\" mission in Pokémon Scarlet and...          0.809698   \n",
       "17  In Pokémon Scarlet and Violet, evolving Eevee ...          0.878603   \n",
       "18  In Pokémon Scarlet and Violet, you can earn mo...          0.883585   \n",
       "19  To effectively tackle the Elite Four in Pokémo...          0.865171   \n",
       "20  Tera Raid Battles in Pokémon Scarlet and Viole...          0.825973   \n",
       "21  In Pokémon Scarlet and Violet, Gimmighoul Coin...          0.913784   \n",
       "22  In Pokémon Scarlet and Violet, obtaining speci...          0.786776   \n",
       "23  Catching legendary Pokémon in Pokémon Scarlet ...          0.809653   \n",
       "24  In Pokémon Scarlet and Violet, the Montenevera...          0.931193   \n",
       "25  In Pokémon Scarlet and Violet, making a sandwi...          0.821579   \n",
       "26  In Pokémon Scarlet and Violet, the real identi...          0.881536   \n",
       "27  The Titan Bombirdier in Pokémon Scarlet and Vi...          0.889183   \n",
       "28  In Pokémon Scarlet and Violet, Area Zero is a ...          0.844975   \n",
       "29  In Pokémon Scarlet and Violet, battling agains...          0.828336   \n",
       "30  In Pokémon Scarlet, evolving Charcadet into Ar...          0.827349   \n",
       "31  In Pokémon Scarlet and Violet, there are a few...          0.851082   \n",
       "32  In Pokémon Scarlet and Violet, during the Cham...          0.865514   \n",
       "33  In Pokémon Scarlet and Violet, Hassel is one o...          0.901105   \n",
       "34  In Pokémon Scarlet and Violet, the Pokémon Lea...          0.855798   \n",
       "35  To turn off the time machine in Area Zero in P...          0.803323   \n",
       "36  In Pokémon Scarlet and Violet, Gimmighoul's Ch...          0.739195   \n",
       "37  In Pokémon Scarlet and Violet, the Soothe Bell...          0.813806   \n",
       "38  If you find yourself going in the wrong direct...          0.839107   \n",
       "39  In Pokémon Scarlet and Violet, you can climb m...          0.834199   \n",
       "40  If you're having trouble getting past Gym 6 in...          0.888818   \n",
       "41  Fighting Dragon-type Pokémon can be challengin...          0.865473   \n",
       "42  In Pokémon Scarlet and Violet, the final battl...          0.823552   \n",
       "43  To get a Polteageist in Pokémon Scarlet and Vi...          0.923085   \n",
       "44  To get a Marill in Pokémon Scarlet and Violet,...           0.86395   \n",
       "45  In Pokémon Scarlet and Violet, there are eight...          0.838411   \n",
       "46  The \"Let's Go\" feature in Pokémon Scarlet and ...          0.896058   \n",
       "47  After you defeat the Elite Four in Pokémon Sca...          0.832983   \n",
       "48  In Pokémon Scarlet and Violet, during the fina...          0.791244   \n",
       "49  In Pokémon Scarlet and Violet, Sunflora is the...          0.922454   \n",
       "\n",
       "   answer_correctness  \n",
       "0            0.689478  \n",
       "1            0.227113  \n",
       "2            0.489086  \n",
       "3            0.365026  \n",
       "4            0.922498  \n",
       "5            0.682897  \n",
       "6            0.220561  \n",
       "7            0.350045  \n",
       "8            0.195525  \n",
       "9            0.429977  \n",
       "10           0.687918  \n",
       "11           0.966497  \n",
       "12           0.448492  \n",
       "13                NaN  \n",
       "14           0.904605  \n",
       "15           0.709207  \n",
       "16           0.923578  \n",
       "17           0.719651  \n",
       "18           0.336257  \n",
       "19           0.216293  \n",
       "20           0.902922  \n",
       "21           0.603446  \n",
       "22           0.339551  \n",
       "23           0.338777  \n",
       "24            0.76221  \n",
       "25           0.955395  \n",
       "26           0.220388  \n",
       "27           0.487001  \n",
       "28           0.863418  \n",
       "29           0.282084  \n",
       "30           0.206837  \n",
       "31            0.77527  \n",
       "32           0.403871  \n",
       "33            0.86164  \n",
       "34           0.642521  \n",
       "35           0.200831  \n",
       "36           0.184799  \n",
       "37           0.203451  \n",
       "38           0.209777  \n",
       "39            0.33355  \n",
       "40           0.395281  \n",
       "41            0.45321  \n",
       "42           0.205888  \n",
       "43           0.605771  \n",
       "44           0.290987  \n",
       "45           0.209603  \n",
       "46           0.599014  \n",
       "47           0.649428  \n",
       "48           0.197811  \n",
       "49           0.313947  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the evaluation results\n",
    "result_df = data[['question', 'ground_truth', 'answer', 'answer_similarity', 'answer_correctness']]\n",
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean Answer Similarity: 0.8589518690737443\n",
      "Mean Answer Correctness: 0.4836676752164233\n"
     ]
    }
   ],
   "source": [
    "# Print the mean values of answer_similarity and answer_correctness\n",
    "print(\"Mean Answer Similarity:\", data['answer_similarity'].replace(np.nan, 0).mean())\n",
    "print(\"Mean Answer Correctness:\", data['answer_correctness'].replace(np.nan, 0).mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
